>**Tip**: Welcome to the Investigate a Dataset project! You will find tips in quoted sections like this to help organize your approach to your investigation. Before submitting your project, it will be a good idea to go back through your report and remove these sections to make the presentation of your work as tidy as possible. First things first, you might want to double-click this Markdown cell and change the title so that it reflects your dataset and investigation.

# Project: No-show Appointments Data Analysis

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>


<a id='intro'></a>
## Introduction

> **Brife Description of the dataset**: In this project I will analyze the Medical Appointment Attendance Rates dataset that have been collected from 100k medical appointments in Brazil. The data has been downloaded as csv file and manipulate using pandas, numpy, matplotlib and jupyter notebook.
     I investigated by breaking down the data categorically to examine common assumptions based on Gender, Age, Economics, Disease, Technical Aptitude, or Wait time were indicators of a patient showing up for an appointment. 
     This dataset is focused on the question of whether or not patients show up for their appointment. A number of characteristics about the patient are included in each row.

> **The questions that can be formulated and answered with the help of this dataset are:**
    1. What is the overall status of patients attendances to appointments?
    2. What are the importance of weekday, hours in attendance of appointments?
    3. What are the importance of age?
    4. What are the importance of health diagnosis like hypertension in attendance of appointments?
    5. What are the importance of health diagnosis like diabetes in attendance of appointments? 
    6. What are the importance of health diagnosis like alcoholism in attendance of appointments?
    7. What are the importance of gender in attendance of appointments?
    8. What are the importance of sms_received in attendance of appointments?


###  Preparing to Investigate

I prepared the data using these common python data analysis tools/packages: numpy, pandas, and matplotlib.

In [31]:
# import the modules 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure

import warnings
warnings.filterwarnings('ignore')
plt.rcParams.update({'font.size': 14})

%matplotlib inline
# I have a retina screen, and this magic makes the charts more sharp
%config InlineBackend.figure_format = 'retina'

<a id='wrangling'></a>
## Data Wrangling

> After observing the dataset and proposed questions for the analysis we will be keeping only relevent data deleting the unsued data so that we can make our calculation easy and understandable.

### General Properties
First, I have to load the data and print out a few lines  Perform operations to inspect data types and look for instances of missing or possibly errant data.

In [32]:
#load the dataset from the csv file into a dataframe
df = pd.read_csv('no-show-dset.csv')
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [33]:
# find number of samples and columns in dataset 
df.shape 

(110527, 14)

In [34]:
# find number of the duplicate rows in dataset
df.duplicated().sum()

0

In [35]:
# find datatypes and all of columns 
df.dtypes

PatientId         float64
AppointmentID       int64
Gender             object
ScheduledDay       object
AppointmentDay     object
Age                 int64
Neighbourhood      object
Scholarship         int64
Hipertension        int64
Diabetes            int64
Alcoholism          int64
Handcap             int64
SMS_received        int64
No-show            object
dtype: object

In [36]:
# find the number of missing values in the dataset
df.isnull().sum()

PatientId         0
AppointmentID     0
Gender            0
ScheduledDay      0
AppointmentDay    0
Age               0
Neighbourhood     0
Scholarship       0
Hipertension      0
Diabetes          0
Alcoholism        0
Handcap           0
SMS_received      0
No-show           0
dtype: int64

In [37]:
# find the number of non-null unique values in the dataset
df.nunique()

PatientId          62299
AppointmentID     110527
Gender                 2
ScheduledDay      103549
AppointmentDay        27
Age                  104
Neighbourhood         81
Scholarship            2
Hipertension           2
Diabetes               2
Alcoholism             2
Handcap                5
SMS_received           2
No-show                2
dtype: int64

In [38]:
# find useful descriptive statistics for each column of data
df.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921784e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hipertension       Diabetes     Alcoholism        Handcap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

        SMS_received  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [39]:
df['No-show'].value_counts()

No     88208
Yes    22319
Name: No-show, dtype: int64

In [40]:
df.head(1)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   

         AppointmentDay  Age    Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62  JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No

### Data Cleaning 
After we assessed the data we have to fix some issue regarding that and do pure clean to come up with readable dataframe

#####  There are multiple issues that need to be fixed in this dataset:

    1. There no extraneous columns that can be dropped
    2. The naming machnisem and spelling is very bad so we need to fix this issue.
    3. The features should be all in lowercase.
    4. The spaces should be replaced it by underscore '_'.
    5. There is no need for filterazation, drop nulls (missing values) or drop duplicates since they are DNE! 
    6. Column 'PatientId' should be converted from float to integer.
    7. Columns 'AppointmentDay' and 'ScheduledDay' should be converted from objects to datetime for consistency.
    8. Column 'Age' also has some erroneous values (-1 and 115).
    9. 'Handcap' has a max value of 4.

In [41]:
# Rename columns to correct spelling errors and for consistency and clarity
df.rename(columns={'PatientId':'Patient Id', 'AppointmentID':'Appointment Id', 'ScheduledDay':'Scheduled Day',
                        'AppointmentDay':'Appointment Day','Neighbourhood':'Location', 'Hipertension':'Hypertension',
                        'Handcap':'Handicap', 'No-show':'No Show', }, inplace=True)
#confirm changes
df.head(1)

Patient Id  Appointment Id Gender         Scheduled Day  \
0  2.987250e+13         5642903      F  2016-04-29T18:38:08Z   

        Appointment Day  Age         Location  Scholarship  Hypertension  \
0  2016-04-29T00:00:00Z   62  JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handicap  SMS_received No Show  
0         0           0         0             0      No

In [42]:
# replace spaces with underscores and lowercase labels
df.rename(columns=lambda x:x.strip().lower().replace(" ", "_"), inplace=True)

#confirm changes
df.head(1)

patient_id  appointment_id gender         scheduled_day  \
0  2.987250e+13         5642903      F  2016-04-29T18:38:08Z   

        appointment_day  age         location  scholarship  hypertension  \
0  2016-04-29T00:00:00Z   62  JARDIM DA PENHA            0             1   

   diabetes  alcoholism  handicap  sms_received no_show  
0         0           0         0             0      No

In [43]:
# convert patient_id from float to int64.
# df['patient_id'].value_counts()
df_v0 = df
df_v0['patient_id'] = df_v0['patient_id'].astype('int64')

#confirm changes
df_v0.head(1)
# df['patient_id'].value_counts()

patient_id  appointment_id gender         scheduled_day  \
0  29872499824296         5642903      F  2016-04-29T18:38:08Z   

        appointment_day  age         location  scholarship  hypertension  \
0  2016-04-29T00:00:00Z   62  JARDIM DA PENHA            0             1   

   diabetes  alcoholism  handicap  sms_received no_show  
0         0           0         0             0      No

In [44]:
#convert scheduled_day and appointment_day from objects to datetime
df_v0['scheduled_day'] = pd.to_datetime(df_v0['scheduled_day'])
df_v0['appointment_day'] = pd.to_datetime(df_v0['appointment_day'])

#confirm changes
df_v0.head(1)

patient_id  appointment_id gender       scheduled_day appointment_day  \
0  29872499824296         5642903      F 2016-04-29 18:38:08      2016-04-29   

   age         location  scholarship  hypertension  diabetes  alcoholism  \
0   62  JARDIM DA PENHA            0             1         0           0   

   handicap  sms_received no_show  
0         0             0      No

In [45]:
df_v0.T

0                    1       \
patient_id            29872499824296      558997776694438   
appointment_id               5642903              5642503   
gender                             F                    M   
scheduled_day    2016-04-29 18:38:08  2016-04-29 16:08:27   
appointment_day  2016-04-29 00:00:00  2016-04-29 00:00:00   
age                               62                   56   
location             JARDIM DA PENHA      JARDIM DA PENHA   
scholarship                        0                    0   
hypertension                       1                    0   
diabetes                           0                    0   
alcoholism                         0                    0   
handicap                           0                    0   
sms_received                       0                    0   
no_show                           No                   No   

                              2                    3       \
patient_id             4262962299951         867951213174   
appointment_id               5642549              5642828   
gender                             F                    F   
scheduled_day    2016-04-29 16:19:04  2016-04-29 17:29:31   
appointment_day  2016-04-29 00:00:00  2016-04-29 00:00:00   
age                               62                    8   
location               MATA DA PRAIA    PONTAL DE CAMBURI   
scholarship                        0                    0   
hypertension                       0                    0   
diabetes                           0                    0   
alcoholism                         0                    0   
handicap                           0                    0   
sms_received                       0                    0   
no_show                           No                   No   

                              4                    5       \
patient_id             8841186448183       95985133231274   
appointment_id               5642494              5626772   
gender                             F                    F   
scheduled_day    2016-04-29 16:07:23  2016-04-27 08:36:51   
appointment_day  2016-04-29 00:00:00  2016-04-29 00:00:00   
age                               56                   76   
location             JARDIM DA PENHA            REPÚBLICA   
scholarship                        0                    0   
hypertension                       1                    1   
diabetes                           1                    0   
alcoholism                         0                    0   
handicap                           0                    0   
sms_received                       0                    0   
no_show                           No                   No   

                              6                    7       \
patient_id           733688164476661        3449833394123   
appointment_id               5630279              5630575   
gender                             F                    F   
scheduled_day    2016-04-27 15:05:12  2016-04-27 15:39:58   
appointment_day  2016-04-29 00:00:00  2016-04-29 00:00:00   
age                               23                   39   
location                  GOIABEIRAS           GOIABEIRAS   
scholarship                        0                    0   
hypertension                       0                    0   
diabetes                           0                    0   
alcoholism                         0                    0   
handicap                           0                    0   
sms_received                       0                    0   
no_show                          Yes                  Yes   

                              8                    9       \
patient_id            56394729949972       78124564369297   
appointment_id               5638447              5629123   
gender                             F                    F   
scheduled_day    2016-04-29 08:02:16  2016-04-27 12:48:25   
appointment_day  2016-04-29 00:00:00  2016-04-29 00:00:00   


In [46]:
df_v0.dtypes

patient_id                  int64
appointment_id              int64
gender                     object
scheduled_day      datetime64[ns]
appointment_day    datetime64[ns]
age                         int64
location                   object
scholarship                 int64
hypertension                int64
diabetes                    int64
alcoholism                  int64
handicap                    int64
sms_received                int64
no_show                    object
dtype: object

While we knew that the age range had unreasonable and inaccurate values, we checked to see how many rows would be affected. The number of rows affected were minimal so the decision was made to remove those rows from the dataframe.

In [47]:
df_v0.loc[(df_v0['age'] < 0) | (df_v0['age'] > 100)].count()

patient_id         8
appointment_id     8
gender             8
scheduled_day      8
appointment_day    8
age                8
location           8
scholarship        8
hypertension       8
diabetes           8
alcoholism         8
handicap           8
sms_received       8
no_show            8
dtype: int64

In [48]:
#Locate rows with age below 0 or above 100.
outAge = df_v0.loc[(df_v0['age'] < 0) | (df_v0['age'] > 100)]

In [49]:
#Drop rows with Age below 0 and above 100.
df_v0.drop(outAge.index, axis=0, inplace=True)

#confirm chnage 
df_v0.loc[(df_v0['age'] < 0) | (df_v0['age'] > 100)].count()

patient_id         0
appointment_id     0
gender             0
scheduled_day      0
appointment_day    0
age                0
location           0
scholarship        0
hypertension       0
diabetes           0
alcoholism         0
handicap           0
sms_received       0
no_show            0
dtype: int64

The featrue Handicap values is not discernable. In the first 4 rows above, it appeared that there was a coorelation between the specified diseases and the values in Handicap. However, the following row deviates from this with a value of 2 and all specified disease values are 0. For this reason, the Handicap column was dropped from the dataframe.

In [50]:
#Check values stored in Handicap and the total of those values.
df_v0['handicap'].value_counts()

0    108282
1      2038
2       183
3        13
4         3
Name: handicap, dtype: int64

In [51]:
#removing Handicap from the dataframe
df_v0.drop(['handicap'], axis=1, inplace=True)

#confirm change
df_v0.head()

patient_id  appointment_id gender       scheduled_day appointment_day  \
0   29872499824296         5642903      F 2016-04-29 18:38:08      2016-04-29   
1  558997776694438         5642503      M 2016-04-29 16:08:27      2016-04-29   
2    4262962299951         5642549      F 2016-04-29 16:19:04      2016-04-29   
3     867951213174         5642828      F 2016-04-29 17:29:31      2016-04-29   
4    8841186448183         5642494      F 2016-04-29 16:07:23      2016-04-29   

   age           location  scholarship  hypertension  diabetes  alcoholism  \
0   62    JARDIM DA PENHA            0             1         0           0   
1   56    JARDIM DA PENHA            0             0         0           0   
2   62      MATA DA PRAIA            0             0         0           0   
3    8  PONTAL DE CAMBURI            0             0         0           0   
4   56    JARDIM DA PENHA            0             1         1           0   

   sms_received no_show  
0             0      No  
1             0      No  
2             0      No  
3             0      No  
4             0      No

##### We could introduce multiple new columns In order to measure the affect of wait-time

An additional column Days was added to calculate the days a patient waited for an appointment.!
    + Add new columns called 'elapsed_days' that take the difference between the 'scheduled_day' and 'appointment_day' to calculate the days a patient waited for an appointment.
    + Add new columns called 'scheduled_dayofweek'
    + Add new columns called 'scheduled_hours'
    + Add new columns called 'appointment_dayofweek'

In [52]:
df_v1 = df_v0
# calculate the total days a patient will wait for their appointment and put that in a new column
df_v1['elapsed_days'] = df_v1['appointment_day'] - df_v1['scheduled_day']

# extract the scheduled days and put it in a new colmun
df_v1['scheduled_dayweek'] = df_v1['scheduled_day'].dt.weekday_name

# extract the scheduled hours and put that in a new column
df_v1['scheduled_hour'] = df_v1['scheduled_day'].apply(lambda time: time.hour)

# extract the appointment days and put it in a new colmun
df_v1['appointment_dayweek'] = df_v1['appointment_day'].dt.weekday_name

# confirm changes 
df_v1.head()

patient_id  appointment_id gender       scheduled_day appointment_day  \
0   29872499824296         5642903      F 2016-04-29 18:38:08      2016-04-29   
1  558997776694438         5642503      M 2016-04-29 16:08:27      2016-04-29   
2    4262962299951         5642549      F 2016-04-29 16:19:04      2016-04-29   
3     867951213174         5642828      F 2016-04-29 17:29:31      2016-04-29   
4    8841186448183         5642494      F 2016-04-29 16:07:23      2016-04-29   

   age           location  scholarship  hypertension  diabetes  alcoholism  \
0   62    JARDIM DA PENHA            0             1         0           0   
1   56    JARDIM DA PENHA            0             0         0           0   
2   62      MATA DA PRAIA            0             0         0           0   
3    8  PONTAL DE CAMBURI            0             0         0           0   
4   56    JARDIM DA PENHA            0             1         1           0   

   sms_received no_show      elapsed_days scheduled_dayweek  scheduled_hour  \
0             0      No -1 days +05:21:52            Friday              18   
1             0      No -1 days +07:51:33            Friday              16   
2             0      No -1 days +07:40:56            Friday              16   
3             0      No -1 days +06:30:29            Friday              17   
4             0      No -1 days +07:52:37            Friday              16   

  appointment_dayweek  
0              Friday  
1              Friday  
2              Friday  
3              Friday  
4              Friday

In [53]:
df_v1.dtypes

patient_id                       int64
appointment_id                   int64
gender                          object
scheduled_day           datetime64[ns]
appointment_day         datetime64[ns]
age                              int64
location                        object
scholarship                      int64
hypertension                     int64
diabetes                         int64
alcoholism                       int64
sms_received                     int64
no_show                         object
elapsed_days           timedelta64[ns]
scheduled_dayweek               object
scheduled_hour                   int64
appointment_dayweek             object
dtype: object

In [54]:
# save final CLEAN dataset as new file
df_v1.to_csv('no_show_dset_clean.csv', index=False)

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. Compute statistics and create visualizations with the goal of addressing the research questions that you posed in the Introduction section. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.

### Research Question 1 (Replace this header name!)

In [4]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.


### Research Question 2  (Replace this header name!)

In [5]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work here, check over your report to make sure that it is satisfies all the areas of the rubric (found on the project submission page at the end of the lesson). You should also probably remove all of the "Tips" like this one so that the presentation is as polished as possible.

## Submitting your Project 

> Before you submit your project, you need to create a .html or .pdf version of this notebook in the workspace here. To do that, run the code cell below. If it worked correctly, you should get a return code of 0, and you should see the generated .html file in the workspace directory (click on the orange Jupyter icon in the upper left).

> Alternatively, you can download this report as .html via the **File** > **Download as** submenu, and then manually upload it into the workspace directory by clicking on the orange Jupyter icon in the upper left, then using the Upload button.

> Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right here. This will create and submit a zip file with this .ipynb doc and the .html or .pdf version you created. Congratulations!

In [6]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'Investigate_a_Dataset.ipynb'])

0